docs/spec/consensus/light-client.md

```diff
@@ -1,113 +1,315 @@
+bisection/recursion. *CheckSupport* implements the trusted period method and, for two adjacent headers (in term of heights), it checks uninterrupted sequence of proof.
+
+*Assumption*: In the following, we assume that *h2.Header.height > h1.Header.height*. We will quickly discuss the other case in the next section.
+
+We consider the following set-up:
+- the lite client communicates with one full node
+- the lite client locally stores all the signed headers it obtained (trusted or not). In the pseudo code below we write *Store(header)* for this.
+- If *Bisection* returns *false*, then the lite client has seen a  forged header.
+  * However, it does not know which header(s) is/are the problematic one(s).
+  * In this case, the lite client can submit (some of) the headers it has seen as evidence. As the lite client communicates with one full node only when executing Bisection, it is safe to assume that this full node is faulty.
+- the lite client must retry to retrieve correct headers from another full node
+  * it picks a new full node
+  * it restarts *Bisection*
+  * there might be optimizations; a lite client may not need to call *Commit(k)*, for a height *k* for which it already has a signed header it trusts.
+
+**Auxiliary Functions.** We will use the  function ```votingpower_in(V1,V2)``` to compute the voting power the validators in set V1 have according to their voting power in set V2;
+we will write ```totalVotingPower(V)``` for ```votingpower_in(V,V)```, which returns the total voting power in V.
+We further use the function ```signers(Commit)``` that returns the set of validators that signed the Commit.
+
+**CheckSupport.** The following function checks whether we can trust the header h2 based on header h1 following the trusting period method.
+
+```go
+  func CheckSupport(h1,h2,trustlevel,trustlevel_adj) bool {
+    if h1.Header.bfttime + tp < now { // Observation 1
+      return false // old header was once trusted but it is expired
+    }  
+    vp_all := totalVotingPower(h1.Header.NextV)
+      // total sum of voting power of validators in h2
+
+    if h2.Header.height == h1.Header.height + 1 {
+      // specific check for adjacent headers; everything must be
+      // properly signed.
+      // also check that h2.Header.V == h1.Header.NextV
+      // Plus the following check that 2/3 of the voting power
+      // in h1 signed h2
+      return (votingpower_in(signers(h2.Commit),h1.Header.NextV) >
+              max(2/3,trustlevel_adj) * vp_all)
+        // signing validators are more than two third in h1.
+        // headers in Cosmos are typically signed by all validators, so one
+        // may request that all signed (liveness might be endangered)
+    }
+
+    s := intersection(h1.Header.NextV,signers(h2.Commit))
+      // get validators in h1 that signed h2
+    vps := votingpower_in(s,h1.Header.NextV)
+      // sum of voting powers in h1 of
+      // validators that signed h2
+    return (vps > max(1/3,trustlevel) * vp_all)
```

Comment:
I feel like we are missing some explanations along the lines of the old ADR.

In theory the lite client should do:
1) get vps, the voting power of the intersection of V1 and V2 in V1
2) substract 1/3rd of total power of V1 (worst case where close to the 1/3rd of potentially faulty validators in V1 are part of the intersection) from vps, be that vps'
3) check vps' > 2/3rd of total power of V2 (guaranteed correct validators in V2 have more than 2/3rds)

The check in 3) fails most of the time, even for small validator set changes. Because of this we have relaxed the check in 3). 
Along these lines, I think trustlevel should be against V2 total power:
```suggestion
    return (vps - 1/3 * vp_all_v1 > trustlevel *vp_all_v2)
```

Reply:
There is a difference to the intersection mechanism in the old version. We do not do intersection and check for two thirds. We check whether at least one *correct* validator from header h1 signed h2. That a correct validator exists, we check by requiring more than 1/3 of the old validators signing h2.

So we rely on the fact that a correct validator only signs "correct" blocks/headers, and that header h2 is well-formed, that is, all the checks are successful.

Reply:
Yes, I am just saying that we should add the justification for the difference.

Reply:
OK. I added a comment on the old method and why we have chosen the new one.